### ISSUE TITLE
Devtools for swift applications

### ISSUE THREAD

**Issue by bolismauro** [zalmoxisus/redux-devtools-extension#288]
Hi everyone,
First of all, this project is amazing :)

I'm one of the creators of a [Swift version of React+Redux](https://github.com/BendingSpoons/katana-lib-swift/) named Katana.

I'm planning to create a dev tools similar to this one but for Katana. 
I was wondering if it is possible to avoid to create a brand new software and use instead this project.
I saw that React Native is supported too, so I guess that is should be doable. What do you think? Can you suggest to me some directions on how to approach the problem?

**Comment by zalmoxisus** [comment]
Hi and thanks for the interest!

Sure, AFAIK, there are integrations already with F#, Python, and Closure. So it's more than possible.

You just have to send data via sockets to [`remotedev-server`](https://github.com/zalmoxisus/remotedev-server) and subscribe to messages (to interpret monitor actions).

I've posted [a draft of the API](https://github.com/zalmoxisus/remotedev-server/blob/v0.3/docs/API/Realtime.md) you can follow. Basically, you need a client for socketcluster. You could adapt the one for [Objective-C](https://github.com/abpopov/SocketCluster-ios-client). Or write a similar one for Swift, which would be much appreciated by the community.

Let me know if you need any further information.

**Comment by bolismauro** [comment]
So basically I'd have
- Browser with the UI
- Remote/Local server that uses `remotedev-server`
- The app written in Swift

The app sends messages to the browser using sockets and the remote/local dev-server is used to a the "master server" for the socket connection. Is this correct?

**Comment by zalmoxisus** [comment]
Exactly. It can interact with the Redux Extension or with other monitor apps listed [here](https://github.com/zalmoxisus/remote-redux-devtools#monitoring). All of them (including this extension) have [`remotedev-app`](https://github.com/zalmoxisus/remotedev-app) as a core lib.

The bridge server ([`remotedev-server`](https://github.com/zalmoxisus/remotedev-server#import-in-your-serverjs-script-you-use-for-starting-a-development-server)) can be started together with Webpack's dev server, from cli, or from the electron app.

**Comment by bolismauro** [comment]
Amazing, I'll take a look as soon as I have some free time :)

**Comment by bolismauro** [comment]
I'm trying to implement this in Objc (I'm using that library for now..)

```
socket.emit('login', 'master', (error, channelName) => {
  if (error) { console.log(error); return; }
  channel = socket.subscribe(channelName);
  channel.watch(handleMessages);
  socket.on(channelName, handleMessages);
});
```

```
    let message = SCMessage.init(eventName: "login", andData: "master".data(using: .utf8))
    message.send(success: {
      print($0)
    }, withFail: {
      print($0)
    })
```

I've the remove-dev server running in my console
```
Test|⇒ node index.js
[RemoteDev] Start server...
--------------------------------------------------------------------------------

   [Busy] Launching SocketCluster
   [Active] SocketCluster started
            Version: 5.2.1
            Environment: dev
            WebSocket engine: uws
            Port: 8000
            Master PID: 82596
            Worker count: 1
            Broker count: 1

   [Done] Migrations are finished
```

But nothing happens.. I've also http://remotedev.io/local/ in a tab in my chrome

**Comment by zalmoxisus** [comment]
As far as I see, you didn't send anything to the monitor yet.

Whether you use http://remotedev.io/local/ or other app, you should click on `Settings button` and check `Use custom local server` (by default it uses `remotedev.io`).

Here's a simple snippet from NodeJS I've tried recently: https://github.com/zalmoxisus/remotedev-server/issues/34#issuecomment-269974168. Unfortunately, I don't know much about  ObjectiveC.

**Comment by zalmoxisus** [comment]
I guess we need more advanced logs for `remotedev-server` so you could see the connected clients and received messages right in the console.

**Comment by bolismauro** [comment]
Uhm yes, that would be useful.
I know for sure that there is a sort of connection because otherwise in the logs of the app I see errors.
Just to make sure I'm doing the right thing: I'm emitting a message "login" with data "master".
Is it correct?

**Comment by zalmoxisus** [comment]
Yes, right. You should get the channel name as a response.

**Comment by bolismauro** [comment]
It seems that the problem is the connection
```
print(self.client.getState() == SOCKET_STATE_CLOSED)
```
prints true (it is evaluated a couple of seconds after `connect`) This is so weird

**Comment by zalmoxisus** [comment]
So, you get that error even with just
```swift
    self.client = SCSocket.client()
    self.client.initWithHost("localhost", onPort: 8000, securely: false)
    self.client.delegate = self
    self.client.connect()
```

Maybe @abpopov or @jondubois could tell if that lib can be used with `socketcluster-server@5`?

**Comment by bolismauro** [comment]
Alright I found the problem
the ws url has `/socketcluster/` in the url and I didn't added it. Actually from the interface I haven't found a way. I've patched the project right now. There are some crashes but I'm solving them.
Let's see

**Comment by bolismauro** [comment]
Success!

For future reference
```swift
let message = SCMessage.init(eventName: "login", andData: ["data": "master"])
```

I'll continue to experiment. If it makes sense to use this monitor I think we will create a swift implementation of ws library
Many thanks for the help!

**Comment by bolismauro** [comment]
@zalmoxisus Your note (without patching the library as I did) leads to this URL
```
ws://localhost/socketcluster/:8000/?transport=websocket
```

Which doesn't work. I haven't found a way to create the proper url without changing the lib

**Comment by zalmoxisus** [comment]
Weird, according to [the example from there](https://github.com/abpopov/SocketCluster-ios-client/blob/ac2d16fa17a671ea8749e15ce1a81ad31e149bbc/Example/SocketCluster-ios-client/SocketClusterClientViewController.m#L28), it should work.

**Comment by bolismauro** [comment]
It may be that I'm doing something wrong..

**Comment by jondubois** [comment]
Personally, I haven't used the native iOS client and I don't manage that repo.

@bolismauro Feel free to fork the repo and makes changes if there are any issues - Then I guess you can make a PR for @abpopov to review.

**Comment by bolismauro** [comment]
Hi.. I didn't have much time to work on this project.. But I finally managed to create a proof of concept :D
https://github.com/bolismauro/katanaMonitor-lib-swift

Thanks for your help guys :)

**Comment by zalmoxisus** [comment]
@bolismauro looks great. Feel free to submit a PR to add it to [the integrations links](https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/Integrations.md).
